rr_iarrc/src/finish_line_watcher/finish_line_watcher.cpp

```diff
@@ -78,30 +78,122 @@ void ImageCB(const sensor_msgs::ImageConstPtr& msg) {
     frame = cv_ptr->image;
     blockEnvironment(frame);
 
-    auto count = countNonZero(frame);
+    // Matrix used to draw debug info
+    cv::Mat debugDrawing = cv::Mat::zeros(frame.size(), CV_8UC3);
+    cv::cvtColor(frame, debugDrawing, cv::COLOR_GRAY2BGR);
+
+    // Find contours
+    vector<vector<cv::Point>> contours;
+    vector<cv::Vec4i> hierarchy;
+    cv::findContours(frame, contours, cv::RETR_LIST, cv::CHAIN_APPROX_SIMPLE);
 
-    auto width = getWidth(frame);
+    // Filter out very small contours
+    vector<cv::Point> filteredContourPoints;
+    for (int i = 0; i < contours.size(); i++) {
+        cv::Scalar color(0, 0, 255);
+
+        if (cv::contourArea(contours.at(i)) > min_contour_area) {
+            filteredContourPoints.insert(std::end(filteredContourPoints), std::begin(contours.at(i)),
+                                         std::end(contours.at(i)));
+            color = cv::Scalar(0, 255, 0);
+        }
+
+        // Debug draw
+        cv::drawContours(debugDrawing, contours, i, color, 2, 8);
+    }
 
-    if (state == LOW && count > 2000 && width > 700) {
+    bool detected = false;
+    bool drawRectDebug = false;
+    cv::RotatedRect fitRect;
+    cv::Size2f size;
+    float angle;
+
+    if (filteredContourPoints.size() > 0) {
+        // Fit a rect to the points and detect based on angle and width
+        fitRect = cv::minAreaRect(filteredContourPoints);
+        size = fitRect.size;
+        angle = std::abs(std::abs(fitRect.angle) - 90);
+        drawRectDebug = true;
+        // Check the detection criteria (must look at both width and height because OpenCV is inconsistent about how it
+        // assigns them)
+        detected = angle < angle_cutoff && (size.width > width_cutoff || size.height > width_cutoff);
+
+        cv::Point2f rectPoints[4];
+        fitRect.points(rectPoints);
+
+        for (int i = 0; i < 4; i++)
+            cv::line(debugDrawing, rectPoints[i], rectPoints[(i + 1) % 4], cv::Scalar(255, 0, 0), 2, 12);
+    }
+
+    // Quick count of pixels as a final sanity check
+    auto count = cv::countNonZero(frame);
+    detected = detected && count > count_thresh;
```

Comment:
Not really sure this is really needed since you are `countNonZero` on the original frame without filtering anyways.

Reply:
Do you mean with the contour stuff?

Reply:
yeah

Reply:
good point I just removed it because I had the threshold uselessly low